internal/controllers/topology/cluster/scope/state.go

```diff
@@ -90,11 +121,44 @@ type MachineDeploymentState struct {
 	MachineHealthCheck *clusterv1.MachineHealthCheck
 }
 
-// IsRollingOut determines if the machine deployment is upgrading.
+// IsRollingOut determines if the machine deployment is rolling out.
 // A machine deployment is considered upgrading if:
 // - if any of the replicas of the machine deployment is not ready.
 func (md *MachineDeploymentState) IsRollingOut() bool {
 	return !mdutil.DeploymentComplete(md.Object, &md.Object.Status) ||
 		*md.Object.Spec.Replicas != md.Object.Status.ReadyReplicas ||
 		md.Object.Status.UnavailableReplicas > 0
 }
+
+// IsUpgrading determines if the MachineDeployment is upgrading.
+// A machine deployment is considered upgrading if at least one of the Machines of this
+// MachineDeployment has a different version.
+func (md *MachineDeploymentState) IsUpgrading(ctx context.Context, c client.Client) (bool, error) {
+	// If the MachineDeployment has no version then we cannot determine if it is upgrading.
+	// Note: This case should not happen.
+	if md.Object.Spec.Template.Spec.Version == nil {
+		return false, nil
+	}
+	clusterName := md.Object.Labels[clusterv1.ClusterNameLabel]
+	mdTopology := md.Object.Labels[clusterv1.ClusterTopologyMachineDeploymentNameLabel]
+	machines := &clusterv1.MachineList{}
```

Comment:
What about using the selector from the MD instead of building our own?

Reply:
This comment is on outdated code. The function is already switched to using the selector from MachineDeployment.